src/SignalR/server/Core/src/DefaultHubLifetimeManager.cs

```diff
@@ -64,6 +74,16 @@ public override Task RemoveFromGroupAsync(string connectionId, string groupName,
             return Task.CompletedTask;
         }
 
+        // Remove from previously saved groups
+        lock (connection.GroupNames)
+        {
+            if (!connection.GroupNames.Remove(groupName))
+            {
+                // Connection not in group
+                return Task.CompletedTask;
+            }
+        }
+
```

Comment:
There is a race here. If `RemoveFromGroupAsync` and `AddToGroupAsync` are called at the same time then we could end up in a state where:
* `GroupNames` contains the group but `_groups` doesn't.
  - I don't think this case results in any "bad" behavior since adding and removing from a group in parallel will already be random as to if you're in the group or not. But it is inconsistent internal state which we'd like to avoid.
* `GroupNames` doesn't contain the group but `_groups` does.
  - This is a little worse because you can't remove the connection from the group until you "add" them again, and on disconnect the connection will forever be in the `_groups` collection.

I think this is easily fixed by moving the `_groups.Remove` and `_groups.Add` calls into the lock.